### ISSUE TITLE
[mono] MAUI release mode crash Assertion at /__w/1/s/src/mono/mono/mini/aot-runtime.c:5220, condition `plt_entry' not met

### ISSUE THREAD

**Issue by jeffgoku** [dotnet/runtime#95406]
### Description

Random crash at the assertion in release mode, no problem for debug mode

### Reproduction Steps

No repro, only happened in release mode

### Expected behavior

should not crash

### Actual behavior

crash

### Regression?

works in dotnet 7

### Configuration

* dotnet 8
* MAUI 8.0.3
* android arm64

**Comment by jonathanpeppers** [comment]
@jeffgoku If you want to try, you can "AOT everything" by setting `AndroidEnableProfiledAot=false` in your `.csproj`.

It's worth checking if that fixes it.

**Comment by jeffgoku** [comment]
> hi @jeffgoku thanks for the bug report. Do you have a sample app that demonstrates the issue that you could share with us?

Sorry I dont have a sample app, I think it will crash after allocate enough memory to trigger a GC

**Comment by jeffgoku** [comment]
> @jeffgoku If you want to try, you can "AOT everything" by setting `AndroidEnableProfiledAot=false` in your `.csproj`.
> 
> It's worth checking if that fixes it.

@jonathanpeppers Thanks, I will give it a try some later time.

**Comment by jurganson** [comment]
@jeffgoku @jonathanpeppers Stumbled upon this, hitting the same issue. Setting AndroidEnableProfiledAot to 'false' seems to have fixed the issue. What is the next move here, is this workaround acceptable? Otherwise, I could use some time to create a small sample.

**Comment by jonathanpeppers** [comment]
If someone could share a `.binlog` of the error, that would help us: https://aka.ms/binlog

In the past, the only case I've seen `AndroidEnableProfiledAot=false` help was if `EnableLLVM=true` was also being used. So, it's worth knowing if that is the case here.

If you are OK with the app size increase, turning off `AndroidEnableProfiledAot` will result in better performance, generally.

**Comment by 4310V343k** [comment]
Could be related: https://github.com/ppy/osu/pull/24500#issuecomment-1676028731

**Comment by anton-yashin** [comment]
Just caught this one:
```
01-22 15:20:27.953 16355 16401 E         : * Assertion at /__w/1/s/src/mono/mono/mini/aot-runtime.c:5220, condition `plt_entry' not met
--------- beginning of crash
01-22 15:20:27.954 16355 16401 F libc    : Fatal signal 6 (SIGABRT), code -6 in tid 16401 (.NET Timer)
01-22 15:20:27.955 16355 16355 I art     : Starting a blocking GC Explicit
01-22 15:20:27.957 16410 16410 I AEE_AED : handle_request(12)
01-22 15:20:27.957 16410 16410 I AEE_AED : check process 16355 name:.myapplication.droid
01-22 15:20:27.957 16410 16410 I AEE_AED : tid 16401 abort msg address:0x0000000000000000 si_code:-6 (request from 16355:10255)
01-22 15:20:27.957 16410 16410 W AEE_AED : debuggerd: handling request: pid=16355 uid=10255 gid=10255 tid=16401
01-22 15:20:27.958 16410 16410 I AEE_AED : [preset_info] pid: 16355, tid: 16401, name: .NET Timer  >>> ru.myapplication.droid <<<
01-22 15:20:28.030 16410 16410 I AEE_AED : *** *** *** *** *** *** *** *** *** *** *** *** *** *** *** ***
01-22 15:20:28.030 16410 16410 I AEE_AED : Build fingerprint: 'Sony/H4311/H4311:7.1.1/49.0.A.6.107/2691393601:user/release-keys'
01-22 15:20:28.030 16410 16410 I AEE_AED : Revision: '0'
01-22 15:20:28.030 16410 16410 I AEE_AED : ABI: 'arm64'
01-22 15:20:28.030 16410 16410 I AEE_AED : pid: 16355, tid: 16401, name: .NET Timer  >>> ru.myapplication.droid <<<
01-22 15:20:28.030 16410 16410 I AEE_AED : signal 6 (SIGABRT), code -6 (SI_TKILL), fault addr --------
01-22 15:20:28.031 16410 16410 I AEE_AED :     x0   0000000000000000  x1   0000000000004011  x2   0000000000000006  x3   0000000000000008
01-22 15:20:28.031 16410 16410 I AEE_AED :     x4   6f635f736600ffff  x5   6f635f736600ffff  x6   000000797d64f000  x7   00000000000000c0
01-22 15:20:28.031 16410 16410 I AEE_AED :     x8   0000000000000083  x9   ffffffffffffffdf  x10  0000000000000000  x11  0000000000000001
01-22 15:20:28.031 16410 16410 I AEE_AED :     x12  0101010101010101  x13  0000000000000010  x14  ffffffffffffffff  x15  000000797a4b7678
01-22 15:20:28.031 16410 16410 I AEE_AED :     x16  000000797a4d3ee0  x17  000000797a47c834  x18  0000000000000000  x19  000000794cfff4f8
01-22 15:20:28.031 16410 16410 I AEE_AED :     x20  0000000000000006  x21  000000794cfff450  x22  0000000000000000  x23  000000794cffeb10
01-22 15:20:28.031 16410 16410 I AEE_AED :     x24  0000000000000001  x25  0000000000000001  x26  000000794cfff030  x27  0000007951736000
01-22 15:20:28.031 16410 16410 I AEE_AED :     x28  000000794cfff050  x29  000000794cffe450  x30  000000797a479c60
01-22 15:20:28.031 16410 16410 I AEE_AED :     sp   000000794cffe430  pc   000000797a47c83c  pstate 0000000060000000
01-22 15:20:28.035 16410 16410 I AEE_AED : 
01-22 15:20:28.035 16410 16410 I AEE_AED : backtrace:
01-22 15:20:28.035 16410 16410 I AEE_AED :     #00 pc 000000000006d83c  /system/lib64/libc.so (tgkill+8)
01-22 15:20:28.035 16410 16410 I AEE_AED :     #01 pc 000000000006ac5c  /system/lib64/libc.so (pthread_kill+64)
01-22 15:20:28.035 16410 16410 I AEE_AED :     #02 pc 0000000000024198  /system/lib64/libc.so (raise+24)
01-22 15:20:28.035 16410 16410 I AEE_AED :     #03 pc 000000000001ca3c  /system/lib64/libc.so (abort+52)
01-22 15:20:28.035 16410 16410 I AEE_AED :     #04 pc 000000000001f360  /data/app/ru.myapplication.droid-1/lib/arm64/libmonodroid.so (_ZN7xamarin7android7Helpers17abort_applicationEv+8)
01-22 15:20:28.035 16410 16410 I AEE_AED :     #05 pc 0000000000035660  /data/app/ru.myapplication.droid-1/lib/arm64/libmonodroid.so (_ZN7xamarin7android8internal16MonodroidRuntime16mono_log_handlerEPKcS4_S4_iPv+144)
01-22 15:20:28.035 16410 16410 I AEE_AED :     #06 pc 00000000001d6fe8  /data/app/ru.myapplication.droid-1/lib/arm64/libmonosgen-2.0.so
01-22 15:20:28.035 16410 16410 I AEE_AED :     #07 pc 00000000001d7114  /data/app/ru.myapplication.droid-1/lib/arm64/libmonosgen-2.0.so
01-22 15:20:28.035 16410 16410 I AEE_AED :     #08 pc 00000000001d7158  /data/app/ru.myapplication.droid-1/lib/arm64/libmonosgen-2.0.so
01-22 15:20:28.035 16410 16410 I AEE_AED :     #09 pc 000000000013f1cc  /data/app/ru.myapplication.droid-1/lib/arm64/libmonosgen-2.0.so
01-22 15:20:28.035 16410 16410 I AEE_AED :     #10 pc 00000000001526b8  /data/app/ru.myapplication.droid-1/lib/arm64/libmonosgen-2.0.so
01-22 15:20:28.035 16410 16410 I AEE_AED :     #11 pc 0000000000004f00  <anonymous:0000007977d25000>
```
Bits from project file
```
    <TargetFramework>net8.0-android</TargetFramework>
    <RuntimeIdentifiers>android-arm;android-arm64</RuntimeIdentifiers>
    <AndroidEnableProfiledAot>True</AndroidEnableProfiledAot>
    <TrimMode>link</TrimMode>
    <EnableLLVM>True</EnableLLVM>
    
      <!-- begin temporary fix for https://github.com/dotnet/runtime/issues/96872 should be removed after february 2024 service release is released -->
  <Target Name="UpdateRuntimePackVersion" BeforeTargets="ProcessFrameworkReferences">
    <ItemGroup>
      <KnownRuntimePack Condition="%(RuntimePackLabels) == 'Mono'" LatestRuntimeFrameworkVersion="8.0.0" />
    </ItemGroup>
  </Target>
  <!-- end temporary fix for https://github.com/dotnet/runtime/issues/96872 should be removed after february 2024 service release is released -->

```
Works fine without EnableLLVM.

**Comment by taublast** [comment]
Enabling trimming for debug didn't change much for me, Debug was running fine.

At the same time, I can confirm that
```
<AndroidEnableProfiledAot>False</AndroidEnableProfiledAot>
<EnableLLVM>True</EnableLLVM>
```
runs okay on `Release`, while
`<EnableLLVM>True</EnableLLVM>`
randomly crashes with `condition 'plt_entry' not met` message.

**Comment by indieshack** [comment]
Can confirm setting AndroidEnableProfiledAot=false stops crashing of my release app caused by "aot-runtime.c:5220, condition `plt_entry' not met", but the APK is now 63mb, twice the size when I don't have that value set to false (and 3x the size of the Xamarin forms APK version the project was migrated from), the compile time on a core i9 is through the roof.

**Comment by jurganson** [comment]
@jonathanpeppers Is the default MAUI-shipped Android AOT profile at cause here? I don't mind the extra AOT time and space requirements but I am simply curious. Also I vaguely remember that creating your own AOT profile in .NET Android was not possible, yet?

**Comment by jonathanpeppers** [comment]
There are some known issues mixing:
```xml
<AndroidEnableProfiledAot>True</AndroidEnableProfiledAot>
<EnableLLVM>True</EnableLLVM>
```

The reason being when we record an AOT profile (recorded under JIT), some of the wrong code paths are taken for checks like `RuntimeFeature.IsSomeHardwareInstrinsticSupported`. I don't think we have a solution for that.

But this combination should work:
```xml
<AndroidEnableProfiledAot>False</AndroidEnableProfiledAot>
<EnableLLVM>True</EnableLLVM>
```
Or this one:
```xml
<!-- Note this is the default if you left these blank -->
<AndroidEnableProfiledAot>True</AndroidEnableProfiledAot>
<EnableLLVM>False</EnableLLVM>
```

You can record your own profile with this NuGet package, but it wouldn't solve this issue:

* https://github.com/jonathanpeppers/Mono.Profiler.Android?tab=readme-ov-file#usage-of-the-aot-profiler

**Comment by veringm** [comment]
I too experience this issue. If one was to choose between 

```
<AndroidEnableProfiledAot>False</AndroidEnableProfiledAot>
<EnableLLVM>True</EnableLLVM>

```
and 

```
<AndroidEnableProfiledAot>True</AndroidEnableProfiledAot>
<EnableLLVM>False</EnableLLVM>
```

I believe AndroidEnableProfiledAot=True is important if app-size is an issue, but the best performance on runtime will be with AndroidEnableProfiledAot=False and EnableLLVM=True. Is that generally correct?

**Comment by beeradmoore** [comment]
If anyone is still looking for a repro of this issue then my [ImageSharpMAUITest](https://github.com/beeradmoore/ImageSharpMAUITest/) (as of commit `ecdf0d92a7195d8f0e9fb74c64cdca4e53db029b`) will reproduce it.

In the csproj I have
```xml
<PropertyGroup Condition="$([MSBuild]::GetTargetPlatformIdentifier('$(TargetFramework)')) == 'android' AND '$(Configuration)' == 'Release'">
    <EnableLLVM>true</EnableLLVM>
    <RunAOTCompilation>true</RunAOTCompilation>
    <AndroidEnableProfiledAot>false</AndroidEnableProfiledAot>
</PropertyGroup>
```

which runs fine. Removing `AndroidEnableProfiledAot` and hitting the `Run all tests` in the application itself will result in a crash. 

> 07-27 14:06:30.804 7895 7961 E gesharpmauitest: * Assertion at /__w/1/s/src/mono/mono/mini/aot-runtime.c:5220, condition `plt_entry' not met
> 07-27 14:06:30.806 7895 7961 F libc : Fatal signal 6 (SIGABRT), code -1 (SI_QUEUE) in tid 7961 (.NET TP Worker), pid 7895 (gesharpmauitest)
 
I did some different things of using different combinations of `EnableLLVM`, `RunAOTCompilation`, and `AndroidEnableProfiledAot` determining performance and filezie. It is written up in [this comment](https://github.com/SixLabors/ImageSharp/pull/2762#issuecomment-2253776267).

The TLDR is,

```xml
<EnableLLVM>false</EnableLLVM>
<AndroidEnableProfiledAot>true</AndroidEnableProfiledAot>
```
Having this setup appears to be the same of just deleting the 2 lines.

---------

```xml
<EnableLLVM>true</EnableLLVM>
<AndroidEnableProfiledAot>false</AndroidEnableProfiledAot>
```
Will give more performance, but at a larger filesize.

---------

```xml
<RunAOTCompilation>true</RunAOTCompilation>
```
May be implied by `EnableLLVM`. Googling "RunAOTCompilation docs" did not lead to any docs on the first page of results, I don't care enough to try hunt ths down on a Saturday to see if that is true or not 😅

**Comment by taublast** [comment]
Make sure you call this one on UI thread. The faster the app works (aot) the more likely your threads execution flow will be differentl from what you had on Debug.

**Comment by ThreeSevenths** [comment]
I also have a repro running into this. https://github.com/ThreeSevenths/MauiSqliteIssue

The workaround posted by [@beeradmoore on 2024-07-27](https://github.com/dotnet/runtime/issues/95406#issuecomment-2253782859) does work for my test app. Although I lost about 5% perf with this combination. 

I think there's an issue with visibility here. `<AndroidEnableProfiledAot>` is enabled by default, and `<EnableLLVM>` is not. If a dev adds `<EnableLLVM>` they could experience this if they like me, naively are working on performance in their apps. No warnings are triggered, and one must identify the correct log entry for the failed assertion to arrive at this GH issue and find a workaround for the problem they are facing. This is time consuming and has many cliffs to fall off and not arrive at this GH issue. The fact that it's been open for more than 18 months does not help. I see this and other issues like it as endemic to the Maui project, and am concerned about the long term viability of any apps developed with it. I hope the team understands that these pitfalls might only affect a small percentage of users from telemetry, but they should still be addressed.

**Comment by Timur-Filatov** [comment]
Hi @jonathanpeppers 
Your contribution to MAUI optimization is difficult to overestimate.

Regarding the current issue, we suggest all MAUI users either to make an app work ~2 times slower with LLVM disabled, but it will start 2 times faster with AOT enabled, or to start 2 times slower, but work ~2 times faster.
Is this the best we can offer (with the love)?

I've converted one of the apps to React Native, it starts 0.5 sec compared to 6 sec for MAUI. Also same lists scrolling without any lags and navigation is 2-3 times faster compared to MAUI CollectionView.
It's not good to ignore this bug in this conditions, even if sometimes MAUI is faster in runtime/logic/bindings.
Thought shadows & swipe view can also be faster on MAUI, but shadows are not working for part of devices due to Android bug (was working in .net8), and some devices will stay without shadows forever. 

My clients complain about low performance of scrolling and navigation on middle Android devices that were bought 1-2 years ago, that it is impossible to work on them because of the lags, I can't offer them to replace the entire fleet of corporate devices with top ones, or to rewrite entire app to RN or anything else.

Currently when using AOT + LLVM without ProfiledAot, apk is 74 MB (AndroidStripILAfterAOT can shrink to 69 MB), there app starts 4 sec, but Azure Blobs not downloading,
If enabling ProfiledAot, Azure Blobs starts to download, but got this "plt_entry not met" crash every 1 minute in random places, the apk is 49 MB, starts 4 sec.

Now disabled AOT at all, app starts 8 sec, but enabled LLVM, so it works faster and apk is 43 MB
(set single arm64 abi per build, because in .net9 there is another bug, if building 2 abis, each apk is 1.5 times larger: 65 MB).

**Comment by jonathanpeppers** [comment]
@Timur-Filatov if there is a specific thing you'd like us to focus on, can you file an issue [here](https://github.com/dotnet/android/issues)?

The conversation here is more about mixing Profiled AOT and LLVM, which doesn't work today.

Are the numbers from app sizes above from Google Play? As you're likely seeing the size of all architectures, densities, languages combined. Google Play knows how to split up app bundles. To get an *estimate* you could build a single architecture to measure app size, like `-r android-arm64`.

**Comment by yasmoradi** [comment]
In iOS, we've option to AOT compile everything, while it fallbacks to Interpreter automatically using the following configuration:
```<MtouchInterpreter>-all</MtouchInterpreter>``` when the currently running code block is not compatible with AOT.
More info at https://learn.microsoft.com/en-us/dotnet/maui/macios/interpreter
Note that this option is not the same as ```<UseInterpreter>true</UseInterpreter>```

In Android, we can AOT compile everything by setting ```<RunAOTCompilation>true</RunAOTCompilation>``` and ```<AndroidEnableProfiledAot>false</AndroidEnableProfiledAot>``` and I *think* it automatically fallbacks to JIT when the running code is not compatible.

What's wrong with ```<EnableLLVM>true</EnableLLVM>```? Why it can't fallback to JIT in case the running code is not compatible with it?

Do we expect things to change regarding to the new Native AOT?

@jonathanpeppers

**Comment by jonathanpeppers** [comment]
NativeAOT doesn't have an interpreter, and you get [build warnings](https://learn.microsoft.com/en-us/dotnet/core/deploying/native-aot/fixing-warnings) for APIs that require it. APIs that require an interpreter won't work at all on NativeAOT.
